Client/Frontend/Home/RemoteTabsPanel.swift

```diff
@@ -13,160 +13,393 @@ import XCGLogger
-        return self.clientAndTabs?[indexPath.section].tabs[indexPath.item]
-    }
-
     override func viewDidLoad() {
         super.viewDidLoad()
+
         tableView.registerClass(TwoLineHeaderFooterView.self, forHeaderFooterViewReuseIdentifier: RemoteClientIdentifier)
         tableView.registerClass(TwoLineTableViewCell.self, forCellReuseIdentifier: RemoteTabIdentifier)
+
         tableView.rowHeight = RemoteTabsPanelUX.RowHeight
         tableView.separatorInset = UIEdgeInsetsZero
 
-        view.backgroundColor = AppConstants.PanelBackgroundColor
+        tableView.delegate = nil
+        tableView.dataSource = nil
 
         refreshControl = UIRefreshControl()
         refreshControl?.addTarget(self, action: "SELrefresh", forControlEvents: UIControlEvents.ValueChanged)
+
+        view.backgroundColor = AppConstants.PanelBackgroundColor
     }
 
     override func viewWillAppear(animated: Bool) {
         super.viewWillAppear(animated)
-        self.SELrefresh()
+        refresh()
     }
 
-    @objc private func SELrefresh() {
-        self.refreshControl?.beginRefreshing()
+    var tableViewDelegate: RemoteTabsPanelDataSource? {
+        didSet {
+            self.tableView.delegate = tableViewDelegate
+            self.tableView.dataSource = tableViewDelegate
+        }
+    }
 
-        self.profile.getClientsAndTabs().upon({ tabs in
-            if let tabs = tabs.successValue {
-                log.info("\(tabs.count) tabs fetched.")
-                self.clientAndTabs = tabs.filter { $0.tabs.count > 0 }
-
-                // Maybe show a background view.
-                let tableView = self.tableView
-                if let clientAndTabs = self.clientAndTabs where clientAndTabs.count > 0 {
-                    tableView.backgroundView = nil
-                    // Show dividing lines.
-                    tableView.separatorStyle = UITableViewCellSeparatorStyle.SingleLine
-                } else {
-                    // TODO: Bug 1144760 - Populate background view with UX-approved content.
-                    tableView.backgroundView = UIView()
-                    tableView.backgroundView?.frame = tableView.frame
-                    tableView.backgroundView?.backgroundColor = UIColor.redColor()
+    func refresh() {
+        tableView.scrollEnabled = false
+        tableView.allowsSelection = false
+        tableView.tableFooterView = UIView(frame: CGRectZero)
+
+        // Short circuit if the user is not logged in
+        if profile.getAccount() == nil {
+            self.tableViewDelegate = RemoteTabsPanelErrorDataSource(homePanel: self, error: .NotLoggedIn)
+            self.tableView.reloadData()
+            return
+        }
 
-                    // Hide dividing lines.
-                    tableView.separatorStyle = UITableViewCellSeparatorStyle.None
+        // Otherwise, fetch the tabs cloud
+        self.profile.getClientsAndTabs().upon({ result in
```

Comment:
I wonder if we document somewhere that this is only safe because `SQLiteRemoteClientAndTabs` has

```
        let deferred = Deferred<Result<[ClientAndTabs]>>(defaultQueue: dispatch_get_main_queue())
```
?